### DISCUSSION TITLE
weed server maintenance

### DISCUSSION THREAD

**Discussion by stallion2303** [seaweedfs/seaweedfs#2005] [General]
Created a stand of 4 VM(3 weed server + 110 replica)

1.  Server `weed server -filer.peers=ip1:8888,ip2:8888,ip3:8888 -master.peers=ip1:9333,ip2:9333,ip3:9333 -dir=/data -dataCenter=DC1 -master.defaultReplication=110 master.volumeSizeLimitMB=1024 -filer rack=rack1 -volume.max=9`
2. Server `weed server -filer.peers=ip1:8888,ip2:8888,ip3:8888 -master.peers=ip1:9333,ip2:9333,ip3:9333 -dir=/data -dataCenter=DC1 -master.defaultReplication=110 master.volumeSizeLimitMB=1024 -filer rack=rack2 -volume.max=9`
3. Server `weed server -filer.peers=ip1:8888,ip2:8888,ip3:8888 -master.peers=ip1:9333,ip2:9333,ip3:9333 -dir=/data -dataCenter=DC2 -master.defaultReplication=110 master.volumeSizeLimitMB=1024 -filer rack=rack1 -volume.max=9`
4. Client `weed mount -filer=ip1:8888 -filer.path=/ -dir=/mnt -cacheCapacityMB=100 -collection=datastore`

The task is to serve one of the servers:

1. ctrl+c on server2 or server3 or both
2. Create a file on the client `echo "test" > /mnt/test`
3. `ls -la /mnt`
-rw-r - r-- 1 root root **5** Apr 15 07:03 test
For the user, the write was successful
4. I see errors in the client's logs
```
/ test saveToStorage [0.5): filerGrpcAddress assign volume: assign volume failure count: 1 collection: "datastore" path: "/ test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection ":" datastore "," replication ": {" rack ": 1," dc ": 1}," ttl ": {" Count ": 0," Unit ": 0}}
E0415 07:03:13 896 filehandle.go: 248] / test doFlush last err: filerGrpcAddress assign volume: assign volume failure count: 1 collection: "datastore" path: "/ test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection": "datastore", "replication": {"rack": 1, "dc": 1}, "ttl": {"Count": 0, "Unit": 0 }}
E0415 07:03:13 896 filehandle.go: 230] Flush doFlush test: input/output error
```
5. `echo "fs.ls -la /" | weed shell`
-rw-r - r-- 0 root root **0** / test
It can be seen that the file was created, but of zero length
6. On the client, ctrl + c, weed mount and `ls -la /mnt`
-rw-r - r-- 1 root root **0** Apr 15 07:03 test
All data was lost.

Questions:
1. How to maintain servers
2. How to switch the cluster to read-only mode
3. How the user understands that the writing is not successful

**Comment by chrislusf** [comment]
If both server2 and server3 are down, the mount should not be able to write. Maybe you need to wait a bit between step 1 and step 2.

**Reply by stallion2303** [reply]
1. shutdown server3
2. sleep 120
3. echo "test" > /mnt/test
4. ls -la /mnt/test
-rw-r--r-- 1 root root **5** Apr 16 14:54 /mnt/test
the write was successful
5. cat /mnt/test
data was lost because output is empty
6. client's log
I0416 07:54:58 12485 dirty_page.go:99] /test saveToStorage [0,5): filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection":"datastore","replication":{"rack":1,"dc":1},"ttl":{"Count":0,"Unit":0}}
E0416 07:54:58 12485 filehandle.go:248] /test doFlush last err: filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection":"datastore","replication":{"rack":1,"dc":1},"ttl":{"Count":0,"Unit":0}}
E0416 07:54:58 12485 filehandle.go:230] Flush doFlush test: input/output error

-  shutdown server2 server3
- sleep 120
- echo "test1" > /mnt/test1
- ls -la /mnt/test1
-rw-r--r-- 1 root root **6** Apr 16 14:56 test1
- cat /mnt/test1
data was lost because output is empty
- client's log
I0416 07:56:54 12485 dirty_page.go:99] /test1 saveToStorage [0,6): filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test1": assign volume: rpc error: code = Unknown desc = raft.Server: Not current leader
E0416 07:56:54 12485 filehandle.go:248] /test1 doFlush last err: filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test1": assign volume: rpc error: code = Unknown desc = raft.Server: Not current leader
E0416 07:56:54 12485 filehandle.go:230] Flush doFlush test1: input/output error

Why doesn't an error occur at the "echo" step?

**Reply by chrislusf** [reply]
There are two operations: creating the file and flushing the content.

Creating the file was successful. But flushing the content failed. The linux pipe seems not checking the close error.

**Reply by stallion2303** [reply]
Single server node down.
`echo "test" | tee /mnt/test`
test
tee: /mnt/test: Input/output error
`ls -la /mnt/test`
-rw-r--r-- 1 root root 5 Apr 19 16:54 /mnt/test

Is it possible to display an error at the operation of file creation and not create a file at all?

Errors appeared after upgrading weed to 2.40. Client's log
I0419 16:54:22 28123 dirty_page.go:99] /test saveToStorage [0,5): filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection":"datastore","replication":{"rack":1,"dc":1},"ttl":{"Count":0,"Unit":0}}
E0419 16:54:22 28123 filehandle.go:246] /test doFlush last err: filerGrpcAddress assign volume: assign volume failure count:1 collection:"datastore" path:"/test": assign volume: rpc error: code = Unknown desc = no free volumes left for {"collection":"datastore","replication":{"rack":1,"dc":1},"ttl":{"Count":0,"Unit":0}}
E0419 16:54:22 28123 filehandle.go:228] Flush doFlush test: input/output error
2021/04/19 16:58:19 fuse: panic in handler for Getxattr [ID=0x15f Node=0x1 Uid=0 Gid=0 Pid=28391] "security.selinux" 255 @0: runtime error: slice bounds out of range [9:1]
goroutine 231 [running]:
github.com/seaweedfs/fuse/fs.(*Server).serve.func2(0x2633aa0, 0xc0000f4000, 0xc002bd9ea0, 0xc002bd9df6)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1019 +0x19e
panic(0x20a6580, 0xc0001306f0)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/runtime/panic.go:965 +0x1b9
github.com/chrislusf/seaweedfs/weed/util/bounded_tree.(*BoundedTree).EnsureVisited(0xc000226270, 0xc0009ec080, 0x1, 0xc002bd99c0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/util/bounded_tree/bounded_tree.go:45 +0x237
github.com/chrislusf/seaweedfs/weed/filesys/meta_cache.EnsureVisited(0xc000032080, 0x2625c50, 0xc00070c750, 0xc0009ec080, 0x1, 0xc0009ec090, 0x9)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/meta_cache/meta_cache_init.go:15 +0x9b
github.com/chrislusf/seaweedfs/weed/filesys.(*WFS).maybeLoadEntry(0xc00070c750, 0xc0009ec080, 0x1, 0xc0009ec081, 0x8, 0x8, 0x1, 0xc003366080)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/xattr.go:117 +0xdc
github.com/chrislusf/seaweedfs/weed/filesys.(*Dir).maybeLoadEntry(0xc0002262a0, 0x9, 0xc003366080, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/dir.go:528 +0x85
github.com/chrislusf/seaweedfs/weed/filesys.(*Dir).Getxattr(0xc0002262a0, 0x2641e58, 0xc002d82040, 0xc0000f4000, 0xc00000cf30, 0x1, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/dir.go:84 +0x96
github.com/seaweedfs/fuse/fs.(*Server).handleRequest(0xc0000fe7e0, 0x2641e58, 0xc002d82040, 0x2606520, 0xc0002262a0, 0xc000033900, 0x2633aa0, 0xc0000f4000, 0xc002bd9ea0, 0x0, ...)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1287 +0x27db
github.com/seaweedfs/fuse/fs.(*Server).serve(0xc0000fe7e0, 0x2633aa0, 0xc0000f4000)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1040 +0x30a
github.com/seaweedfs/fuse/fs.(*Server).Serve.func1(0x5d)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:512 +0x79
github.com/seaweedfs/fuse/fs.(*Server).Serve.func2(0xc0000fe7e0, 0xc000226930, 0x5d)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:519 +0x64
created by github.com/seaweedfs/fuse/fs.(*Server).Serve
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:517 +0x385
2021/04/19 16:58:38 fuse: panic in handler for Getxattr [ID=0x1b6 Node=0x1 Uid=0 Gid=0 Pid=28401] "security.selinux" 255 @0: runtime error: slice bounds out of range [9:1]
goroutine 190 [running]:
github.com/seaweedfs/fuse/fs.(*Server).serve.func2(0x2633aa0, 0xc002df1b30, 0xc002bc3ea0, 0xc002bc3df6)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1019 +0x19e
panic(0x20a6580, 0xc000053308)
	/home/travis/.gimme/versions/go1.16.3.linux.amd64/src/runtime/panic.go:965 +0x1b9
github.com/chrislusf/seaweedfs/weed/util/bounded_tree.(*BoundedTree).EnsureVisited(0xc000226270, 0xc002d1c200, 0x1, 0xc002bc39c0, 0x0, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/util/bounded_tree/bounded_tree.go:45 +0x237
github.com/chrislusf/seaweedfs/weed/filesys/meta_cache.EnsureVisited(0xc000032080, 0x2625c50, 0xc00070c750, 0xc002d1c200, 0x1, 0xc002d1c210, 0x9)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/meta_cache/meta_cache_init.go:15 +0x9b
github.com/chrislusf/seaweedfs/weed/filesys.(*WFS).maybeLoadEntry(0xc00070c750, 0xc002d1c200, 0x1, 0xc002d1c201, 0x8, 0x8, 0x1, 0xc0001d7400)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/xattr.go:117 +0xdc
github.com/chrislusf/seaweedfs/weed/filesys.(*Dir).maybeLoadEntry(0xc0002262a0, 0x9, 0xc0001d7400, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/dir.go:528 +0x85
github.com/chrislusf/seaweedfs/weed/filesys.(*Dir).Getxattr(0xc0002262a0, 0x2641e58, 0xc002d0e340, 0xc002df1b30, 0xc003488390, 0x1, 0x0)
	/home/travis/gopath/src/github.com/chrislusf/seaweedfs/weed/filesys/dir.go:84 +0x96
github.com/seaweedfs/fuse/fs.(*Server).handleRequest(0xc0000fe7e0, 0x2641e58, 0xc002d0e340, 0x2606520, 0xc0002262a0, 0xc000033900, 0x2633aa0, 0xc002df1b30, 0xc002bc3ea0, 0x0, ...)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1287 +0x27db
github.com/seaweedfs/fuse/fs.(*Server).serve(0xc0000fe7e0, 0x2633aa0, 0xc002df1b30)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:1040 +0x30a
github.com/seaweedfs/fuse/fs.(*Server).Serve.func1(0x34)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:512 +0x79
github.com/seaweedfs/fuse/fs.(*Server).Serve.func2(0xc0000fe7e0, 0xc000226930, 0x34)
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:519 +0x64
created by github.com/seaweedfs/fuse/fs.(*Server).Serve
	/home/travis/gopath/pkg/mod/github.com/seaweedfs/fuse@v1.1.4/fs/serve.go:517 +0x385

**Reply by chrislusf** [reply]
Added a fix https://github.com/chrislusf/seaweedfs/commit/83cf94ad2dea74e5f9a881a617d06c42cbb7ba49
